### DISCUSSION TITLE
SSIM Alternative: DISTS

### DISCUSSION THREAD

**Discussion by torridgristle** [eps696/aphantasia#15] [Ideas]
I'm trying [DISTS](https://github.com/dingkeyan93/DISTS) as an alternative to SSIM and so far it's... It works to make the supplied image show up in the results. I don't know a whole lot about it though, to be frank it's just something different that works. I haven't decided if it's of any worth yet.

To get it running in Aphantasia I just needed to `!pip install dists-pytorch` and in clip_fft.py add `from DISTS_pytorch import DISTS` and replace `ssim_loss = ssim.SSIM(window_size = 11)` with `ssim_loss = DISTS(require_grad=True, batch_average=True)`

The two images to compare might need to be blurred before comparison to focus more on shapes.

**Comment by eps696** [comment]
thanks for info, let me know if you find it better than SSIM (i'm not planning to investigate such details soon).

good point about blurring as well; it should probably be done before applying that structure loss. i believe that CLIP comparison must be done with original image to keep all important features intact.

**Comment by torridgristle** [comment]
When I set it to recreate an image using just SSIM, no encoding whatsoever, it tends to look really smoothed out but the edges are there. However, when I tried it with DISTS it let all these neat shapes remain.
![jerma2-512](https://user-images.githubusercontent.com/23273466/112204738-651dec00-8bea-11eb-9149-4b2c4e1a2f5a.jpg)
![jermarecreated](https://user-images.githubusercontent.com/23273466/112204067-a06beb00-8be9-11eb-81ef-f23a82f4ca14.jpg)

Also, it turns out earlier it wasn't even using clip_fft.py in the notebook, I had to put it in the Generate code cell and figure out why it kept complaining about all the tensors not being on the same device, only to realize that the loss apparently wasn't.

To actually get it working, had to add `from DISTS_pytorch import DISTS` and change `ssim_loss = ssim.SSIM(window_size = 11)` to `ssim_loss = DISTS().cuda()` and change `ssim_loss(img_out, img_in)` to `ssim_loss(img_out, img_in, require_grad=True, batch_average=True)`, though it started off giving higher values than expected, like 0.55 or so, so I decided to square the output of it and also changed how it's used to calculate loss for testing, with
```
ssim_loss_output = ssim_loss(img_out, img_in, require_grad=True, batch_average=True)**2.0
loss = loss * (1.0 - sync) + ssim_loss_output * sync
```
which, at least for this, works well.

It doesn't overly match the image, but the overall general shapes are familiar.
![download (3)](https://user-images.githubusercontent.com/23273466/112205319-0efd7880-8beb-11eb-9460-cdc2b23ae867.jpg)
![download (5)](https://user-images.githubusercontent.com/23273466/112206134-f6da2900-8beb-11eb-87a5-d058b511a311.jpg)
with a sync of 0.5 that doesn't change as steps increase.

And I've been experimenting with a touch of weight decay in Adam in an attempt to break up big areas of nothing. Doesn't seem to help with that, doesn't seem to hurt, but it does make it wiggle and jiggle a bit in the final video which is really dreamlike, especially when the loss is turned up.

[Aphantasia_DISTS_Test.zip](https://github.com/eps696/aphantasia/files/6192189/Aphantasia_DISTS_Test.zip)
Here's my Colab notebook with it all set up. Had to manually !wget the weights for DISTS and move it, somehow the install didn't handle that on its own.

**Comment by eps696** [comment]
i tested DISTS and didn't find it convincing. besides quite fuzzy output, it tends to eat GPU RAM way more than SSIM. 
what i used instead, was your idea about blurring: simple 8x downsampling before SSIM removed sharp edges and delivered exactly that level of details which i've been seeking (see below). so thanks for putting me up to this, and best of luck with further investigations. 

text input = "winter", 200 steps, 240 samples, image input:
![rfff025eet](https://user-images.githubusercontent.com/4808408/112233748-05582d00-8c4c-11eb-80ff-05d2603f89d3.png)

sync = 0.5
![rfff025eet-winter-c0 5-200](https://user-images.githubusercontent.com/4808408/112233165-f624af80-8c4a-11eb-80c8-97f5c18463e7.jpg)

sync = 1.0
![rfff025eet-winter-c1-200](https://user-images.githubusercontent.com/4808408/112233679-e5286e00-8c4b-11eb-96de-ff79adc2da68.jpg)

**Comment by russelldc** [comment]
@eps696 Not sure if you noticed, but this SSIM change (8x downsample) introduced a grid of tiny white or black pixels across the whole image. I was going to share my own example, but if you look closely at the 2 images you just posted, you should be able to see it.

**Comment by eps696** [comment]
sure noticed, they go with time (200-300 iters was enough for me)

**Comment by russelldc** [comment]
Odd, the grid doesn't seem to ever fade away in my experience. I've done several 1000+ iteration generations this afternoon, and it seems like it's a permanent effect.

**Comment by eps696** [comment]
well, few dots stay permanent indeed (like on the pics above), but their grid disappeared pretty quickly on all my tests. so i neglected this in favour of kinda proper reconstruction (any other trick didn't make it). 
[i tried to fix it with additional blurring, but then details and edges of original image started appearing again, so i kept it like this]

**Comment by eps696** [comment]
just to fix the issue here: SSIM loss is now replaced by LPIPS, which allows way more delicate image integration, doesn't change colors (like SSIM), doesn't introduce ugly artefacts (like DISTS), and doesn't require downsampling (which introduced b/w dots). hopefully it will stay like that.
@torridgristle @russelldc
